swig/python/codegen/codegen.py

```diff
@@ -0,0 +1,1172 @@
+        """CodeGenerator(tree, file=sys.stdout) -> None.
+         Print the source for tree to file."""
+        self.f = file
+        self.future_imports = []
+        self._indent = 0
+        # dict of locals used to determine if variable already exists in assignments
+        self._locals = ["pyflamegpu"]
+        self._device_functions = []
+        self._message_iterator_var = None           # default
+        self._input_message_var = 'message_in'      # default
+        self._output_message_var = 'message_out'    # default
+        self.dispatch(tree)
+        print("", file=self.f)
+        self.f.flush()
+        
+                
+    def _deviceVariableFunctionName(self, tree, permitted_prefixes, allow_lengths = True):
+        """
+        Gets the device function name by translating a typed Python version to a templated cpp version.
+        Python functions looks like getVariableFloatArray6 and translate to getVariable<float, 6>
+        This function will detect and test against a set of known types and also extract the Array length
+        This function returns None if the string is invalid in format but only throws an error if the format is correct but the type is invalid.
+        """
+        cpp_func_name = ""
+        py_func = tree.attr
+        # extract function name start
+        for prefix in permitted_prefixes:
+            if py_func.startswith(prefix):
+                cpp_func_name = prefix
+                py_func = py_func[len(prefix):]
+                break # dont allow the else
+        else:
+            return None
+        # check type and lengths
+        if allow_lengths:
+            #split to get type and Array Length (This could **potentially** be looked up from the model description but current syntax is consistent with swig bindings)     
+            type_and_length = py_func.split("Array")
+            if type_and_length[0] not in self._fgpu_types:
+                self.RaiseError(tree, f"'{type_and_length[0]}' is not a valid FLAME GPU type")
+            t = self._fgpu_types[type_and_length[0]]
+            # generate template args
+            if (len(type_and_length) == 1):
+                cpp_func_name += f"<{t}>"
+            elif (len(type_and_length) == 2):
+                cpp_func_name += f"<{t}, {type_and_length[1]}>"
+            else:
+                return None
+        else:
+            if py_func not in self._fgpu_types:
+                self.RaiseError(tree, f"'{py_func}' is not a valid FLAME GPU type")
+            t = self._fgpu_types[py_func]
+            cpp_func_name += f"<{t}>"
+        # return    
+        return cpp_func_name
+              
+
+    def fill(self, text = ""):
+        "Indent a piece of text, according to the current indentation level"
+        self.f.write("\n"+"    "*self._indent + text)
+
+    def write(self, text):
+        "Append a piece of text to the current line."
+        self.f.write(six.text_type(text))
```

Comment:
```suggestion
        self.f.write(str(text))
```

`six.text_type` is just `str` in python3.